mozetl/maudau/maudau.py

```diff
@@ -0,0 +1,221 @@
+from moztelemetry.standards import (
+    filter_date_range,
+    count_distinct_clientids
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+DEVELOPMENT = False
+
+STORAGE_BUCKET = "net-mozaws-prod-us-west-2-pipeline-analysis"
+STORAGE_SUB_DIR = 'spenrose/maudau' if DEVELOPMENT else 'mreid/maudau'
+MAUDAU_ROLLUP_BASENAME = "engagement_ratio.csv"
+MAUDAU_SNAPSHOT_TEMPLATE = "engagement_ratio.{}.csv"
+DASHBOARD_BUCKET = "net-mozaws-prod-metrics-data"
+DASHBOARD_KEY = "firefox-dashboard/{}".format(MAUDAU_ROLLUP_BASENAME)
+
+ACTIVITY_SUBMISSION_LAG = DT.timedelta(10)
+# MONTH as in "monthly active"
+MONTH = 28
+# Recency in days of data which should be re-counted.
+UNSTABLE_WINDOW = 30
+
+
+def format_as_submission_date(date):
+    return DT.date.strftime(date, "%Y%m%d")
+
+
+def format_spark_path(bucket, prefix):
+    return "s3://{}/{}".format(bucket, prefix)
+
+
+def get_rollup_s3_paths(basename):
+    return (STORAGE_BUCKET, os.path.join(STORAGE_SUB_DIR, basename))
+
+
+def generate_filter_parameters(end_date, days_back):
+    d = {}
+    min_activity_date = end_date - DT.timedelta(days_back)
+    d['min_activity_iso'] = min_activity_date.isoformat()
+    d['max_activity_iso'] = (end_date + DT.timedelta(1)).isoformat()
+
+    d['min_submission_string'] = format_as_submission_date(min_activity_date)
+    max_submission_date = end_date + ACTIVITY_SUBMISSION_LAG
+    d['max_submission_string'] = format_as_submission_date(max_submission_date)
+    return d
+
+
+def count_active_users(frame, end_date, days_back):
+    """
+    Tally active users according to externally-defined heuristics,
+    specifically client-reported subsession_start_date.
+
+    :frame DataFrame(): conforming to main_summary schema
+    :end_date DT.date(): the last day on which to count activity.
+    :days_back int: how far back to go; 0 for daily and 28 for monthly.
+
+    Note that return values are not stable over time as activity trickles
+    in with long submission latencies.
+
+    See https://bugzilla.mozilla.org/show_bug.cgi?id=1240849
+    """
+    params = generate_filter_parameters(end_date, days_back)
+
+    filtered = filter_date_range(
+        frame,
+        frame.subsession_start_date,
+        params['min_activity_iso'],
+        params['max_activity_iso'],
+        frame.submission_date_s3,
+        params['min_submission_string'],
+        params['max_submission_string'])
+    return count_distinct_clientids(filtered)
+
+
+def parse_last_rollup(basename, start_date=None):
+    """
+    Extract the valid old counts from basename and return them with the
+    first date that needs to be re-counted.
+    """
+    start_date = start_date or DT.date.today()
+    since = (start_date - DT.timedelta(UNSTABLE_WINDOW))
+    since = format_as_submission_date(since)
+    carryover = []
+    with open(basename) as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            if row['day'] >= since:
```

Comment:
Hm, at the very least, we should avoid re-computing all 30 recent days every day and skip anything that accounts for the `ACTIVITY_SUBMISSION_LAG` period. The extra 20 days of processing would add a non-trivial amount of run time (if I recall correctly).

I agree that using a CSV file as the source of truth here is... suboptimal to say the least. We should address that general problem in a follow-up though, as it will probably require significant refactoring and testing to make it both idempotent and incremental.